guja-core/src/main/java/com/wadpam/guja/cache/CacheMethodInterceptor.java

```diff
@@ -0,0 +1,188 @@
+import org.aopalliance.intercept.MethodInvocation;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Provider;
+import java.io.Serializable;
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Method interceptor for adding caching to core crud methods.
+ *
+ * @author sosandstrom
+ * @author mattiaslevin
+ */
+public class CacheMethodInterceptor implements MethodInterceptor {
+  static final Logger LOGGER = LoggerFactory.getLogger(CacheMethodInterceptor.class);
+
+  private final Map<String, Cache<Triple, Optional<?>>> namespaces = new HashMap<>();
+
+  private final Provider<CacheBuilder> cacheBuilderProvider;
+  public CacheMethodInterceptor(Provider<CacheBuilder> cacheBuilderProvider) {
+    this.cacheBuilderProvider = cacheBuilderProvider;
+  }
+
+  private static final String GET_METHOD_NAME = "get";
+  private static final String PUT_METHOD_NAME = "put";
+  private static final String DELETE_METHOD_NAME = "delete";
+
+  static Method readMethod, putMethod, deleteMethod, pageMethod;
+
+  static {
+    try {
+      // TODO read, put, delete does not work with equals, not sure why
+      //readMethod = CrudDao.class.getMethod("get", Serializable.class);
+      //putMethod = CrudDao.class.getMethod("put", Object.class);
+      //deleteMethod = CrudDao.class.getMethod("delete", Serializable.class);
+      pageMethod = CrudDao.class.getMethod("queryPage", Object.class, int.class, String.class);
+    } catch (NoSuchMethodException e) {
+      e.printStackTrace();
+    }
+  }
+
+  @Override
+  public Object invoke(final MethodInvocation invocation) throws Throwable {
+
+    final Class clazz = getClass(invocation);
+    Cache<Triple, Optional<?>> cache = getCacheInstance(clazz);
+
+    final Object[] args = invocation.getArguments();
+    final Triple triple = new Triple(args);
+    final Method method = invocation.getMethod();
+    LOGGER.trace("invoking on {}, isAnnotated {}", method, method.isAnnotationPresent(Cached.class));
+
+    if (PUT_METHOD_NAME.equals(method.getName()) && method.isAnnotationPresent(Crud.class)) {
+      LOGGER.trace("   put");
+      final Object id = invocation.proceed();
+      args[1] = id;
+      final Object entity = args[2];
+      args[2] = null;
+      checkNotNull(id);
+      checkNotNull(entity);
+      cache.put(new Triple(args), Optional.of(entity));
+      return id;
+    }
+
+    if (DELETE_METHOD_NAME.equals(method.getName()) && method.isAnnotationPresent(Crud.class)) {
+      LOGGER.info("   delete");
+      invocation.proceed();
+      final Object id = args[1];
+      checkNotNull(id);
+      cache.put(triple, Optional.absent());
+      return null;
+    }
+
+    final Optional<?> optionalEntity = cache.get(triple, new Callable<Optional<?>>() {
```

Comment:
I guess this flow will handle both get and page read.
Do we need to kind of check that this is actually the case? If someone annotates some other method by misstake.